### DISCUSSION TITLE
pallet-bounties | Sub-bounties Extn | Design Discussion Notes

### DISCUSSION THREAD

**Discussion by shamb0** [paritytech/substrate#7959] [General]
# pallet-bounties | Sub-bounties Extn | Design Discussion Notes

## Overview

![](https://i.imgur.com/8LluAr5.png)

## API & Types

* Trait
	- type MaxSubBountyCount: u32
* Types    
    * struct SubBounty **[New Struct]**
        - proposer: AccountId,
        - value: Balance,
        - curator: AccountId,
        - fee: Balance,
        - curator_deposit: Balance,
        - bond: Balance,
        - status: BountyStatus,
    * struct Bounty **[Existing Struct Extn]**
        - proposer: AccountId,
        - value: Balance,
        - fee: Balance,
        - curator_deposit: Balance,
        - bond: Balance,
        - status: BountyStatus,
        - **subbountycount: BountyIndex,**    
        - **activesubbounty: Vec<BountyIndex>,**
* Error Code **[Extn to Existing codes]**
    - InsufficientBountyBalance,
    - SubBountyActive,
    - SubBountyMaxReached,
* Event **[Extn to Existing events]**
    - SubBountyApproved(BountyIndex, BountyIndex),
    - SubBountyAwarded(BountyIndex, BountyIndex, AccountId),
    - SubBountyClaimed(BountyIndex, BountyIndex, Balance, AccountId),
    - SubBountyRejected(BountyIndex, BountyIndex, Balance),
    - SubBountyCanceled(BountyIndex, BountyIndex),
    - SubBountyBecameActive(BountyIndex, BountyIndex),		
	- SubBountyExtended(BountyIndex, BountyIndex),
* Storage **[Extn to Existing DB fields]**
    - SubBounties: double_map(bounty_id,subbounty_id) => Option<SubBounty>,
    - SubBountyDescriptions: double_map(bounty_id,subbounty_id) => Option<Vec<u8>>,
    - SubBountyApprovals: Vec<(bounty_id,subbounty_id)>,
* Modules **[Extn to Existing Extrincis]**
    - add_subbounty(origin, bounty_id: BountyIndex, value: BalanceOf<T>, description: Vec<u8>);
    - propose_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, subcurator: <T::Lookup as StaticLookup>::Source, fee: BalanceOf<T>);
    - accept_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
    - unassign_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
    - award_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, beneficiary: <T::Lookup as StaticLookup>::Source);
    - claim_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
    - close_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
    - extend_subbounty_expiry(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, remark: Vec<u8>);
    
---
## Quick Proto version

Plz find proto version of the implementation | https://github.com/paritytech/substrate/pull/7965

* Proposed data types and api signatures may not be in right shape, may require review & cleanup.

* Grey part, where may need additional clarity, are marked with below tag. & current logic implemented based on blind assumption.

```
// TODO :: Have to recheck
```

---

## DB

![](https://i.imgur.com/Y08E5xY.png)

---
## State

![](https://i.imgur.com/0theHKh.png)

---

### Reference discussion threads

* [Treasury-pallet Bounty Extension · Issue #5713 · paritytech/substrate](https://github.com/paritytech/substrate/issues/5713)

* [Post #86 | Reopening the discussion on Bounties and Treasury spending issues! Open questions](https://kusama.polkassembly.io/post/86)

* [Post #237 | A Final Check on Bounties' Development](https://kusama.polkassembly.io/post/237)

* [Post #151 | On Bounties Spending parameters - A Summary](https://kusama.polkassembly.io/post/151)

**Comment by xlc** [comment]
* Types
  * struct SubBounty
    * proposer: AccountId,
      _* Why do we need this?_
    * value: Balance,
      * _Why do we need this? This is required for bounty for the funding phase._
      _* But there isn't any funding phase for sub-bounty and you can always query the sub-bounty account holding for the value_
    * curator: AccountId,
    * fee: Balance,
    * curator_deposit: Balance,
    * bond: Balance,
      _* No need to have this._
      _* This is required for bounty because anyone can propose bounty and save data into BountyDescription._
      _* But only approved curator can create sub-bounty which already have a slashable deposit so I don't think we will need_  BountyDescription deposit for sub-bounty
    * status: BountyStatus,
      _* Should have enum SubBountyStatus_
  * struct Bounty
    * proposer: AccountId,
    * value: Balance,
    * fee: Balance,
    * curator_deposit: Balance,
    * bond: Balance,
    * status: BountyStatus,
    * **subbountycount: BountyIndex,**
      _* Do we need this? We can use `BountyCount` to generate sub-bounty id as well_
    * **activesubbounty: Vec,**
      _* Do we need this? It is possible to iterate `SubBounties` by `bounty_id` to fetch all sub-bounties for a particular bounty_
* Storage **[Extn to Existing DB fields]**
  * SubBounties: double_map(bounty_id,subbounty_id) => Option\<SubBounty>,
  * SubBountyDescriptions: double_map(bounty_id,subbounty_id) => Option<Vec>,
    _* No need to have this. Just reuse BountyDescriptions_
  * SubBountyApprovals: Vec<(bounty_id,subbounty_id)>,
    _* There is no approval / funding phase for subbounties so no need to have this_
* Modules **[Extn to Existing Extrincis]**
  * add_subbounty(origin, bounty_id: BountyIndex, value: BalanceOf, description: Vec);
  * propose_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, subcurator: <T::Lookup as StaticLookup>::Source, fee: BalanceOf);
  * accept_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
  * unassign_subcurator(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
  * award_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, beneficiary: <T::Lookup as StaticLookup>::Source);
  * claim_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
  * close_subbounty(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex);
  * extend_subbounty_expiry(origin, bounty_id: BountyIndex, subbounty_id: BountyIndex, remark: Vec);
    _* No need to have this as the expiry for sub-bounties are shared with the parent_

To prevent abuse of BountyDescriptions by create large amount of sub-bounties with large BountyDescriptions for free, we should have a `MaxSubBountyCount`. We can have a `activeSubBountyCount` in `struct Bounty` to enforce that.

**Reply by shamb0** [reply]
Thanks Bryan, Good one. May be I shall make MaxSubBountyCount part of trait config.

**Comment by smisty** [comment]


**Reply by shamb0** [reply]
> What are you people doing to me
> […](#)

Hi @smisty, its discussion thread, capturing some design input for pallet-bounties extn.
Not sure how it disturbed you. Could be accident, plz kindly ignore ...
